### PR TITLE
Add in ability to encrypt/decrypt msg with keyring

### PR DESCRIPTION
**Pull request by RoyTimes**
Ref: https://github.com/polkadot-js/common/issues/1068

- 1 dependency added to 'util-crypto' - `ed2curve` by the author of tweetnacl.js 
- tests included
- covers sr25519 & ed25519, secp256k1 to be added.

### CODE REVIEW

File: packages/util-crypto/src/nacl/convertKey.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Copyright 2017-2021 @polkadot/util-crypto authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+/* eslint-disable */
```
Comment (jacogr) [inline packages/util-crypto/src/nacl/convertKey.ts:4]: Would not do this at all. Rather add a small `ed2curve.d.ts`

File: packages/keyring/src/pair/index.ts [hunk 2]
```diff
@@ -145,6 +145,21 @@ export function createPair ({ toSS58, type }: Setup, { publicKey, secretKey }: P
...
+      const _senderPublicKey = convertPublicKeyToCurve25519(u8aToU8a(senderPublicKey));
+
+      const message = naclOpen(
+        u8aToU8a(encryptedMessageWithNonce.slice(24, encryptedMessageWithNonce.length)),
+        u8aToU8a(encryptedMessageWithNonce.slice(0, 24)),
+        _senderPublicKey,
+        _secretKey
+      );
+
+      return message;
```
Comment (jacogr) [inline packages/keyring/src/pair/index.ts:161]:
```suggestion
      assert(!isLocked(secretKey), 'Cannot encrypt with a locked key pair');

      return naclOpen(
        u8aToU8a(encryptedMessageWithNonce.slice(24, encryptedMessageWithNonce.length)),
        u8aToU8a(encryptedMessageWithNonce.slice(0, 24)),
        convertPublicKeyToCurve25519(u8aToU8a(senderPublicKey)),
        convertSecretKeyToCurve25519(secretKey)
      );

      return message;
```

File: packages/keyring/src/pair/index.ts [hunk 3]
```diff
@@ -157,6 +172,16 @@ export function createPair ({ toSS58, type }: Setup, { publicKey, secretKey }: P
...
     },
+    encryptMessage: (message: string | Uint8Array, recipientPublicKey: string | Uint8Array, _nonce?: Uint8Array): Uint8Array => {
+      assert(!isLocked(secretKey), 'Cannot encrypt with a locked key pair');
+
+      const _secretKey = convertSecretKeyToCurve25519(secretKey);
+      const _recipientPublicKey = convertPublicKeyToCurve25519(u8aToU8a(recipientPublicKey));
+
+      const { nonce, sealed } = naclSeal(u8aToU8a(message), _secretKey, _recipientPublicKey, _nonce);
+
+      return u8aConcat(nonce, sealed);
```
Comment (jacogr) [inline packages/keyring/src/pair/index.ts:183]:
```suggestion
      assert(!isLocked(secretKey), 'Cannot encrypt with a locked key pair');

      const { nonce, sealed } = naclSeal(u8aToU8a(message), convertSecretKeyToCurve25519(secretKey), convertPublicKeyToCurve25519(u8aToU8a(recipientPublicKey)), _nonce);

      return u8aConcat(nonce, sealed);
```

File: packages/util-crypto/src/nacl/index.ts [hunk 4]
```diff
@@ -15,3 +15,7 @@ export { naclVerify } from './verify';
 export { naclBoxKeypairFromSecret } from './box/fromSecret';
 export { naclOpen } from './open';
 export { naclSeal } from './seal';
+export {
+  convertSecretKeyToCurve25519,
+  convertPublicKeyToCurve25519
+} from './convertKey';
```
Comment (jacogr) [inline packages/util-crypto/src/nacl/index.ts:21]:
```suggestion
export { convertSecretKeyToCurve25519, convertPublicKeyToCurve25519 } from './convertKey';
```

**Review by jacogr — COMMENTED** [review]
Funky hack to make this operational in sr/ed/eth.

Some small comments.

The only think I feel is missing is possibly explicit tests for various keypair types -

- Ethereum
- sr25519
- ed25519

File: packages/util-crypto/src/nacl/convertKey.ts (hunk 1 above)
Comment (RoyTimes) [reply packages/util-crypto/src/nacl/convertKey.ts:4]: nvm. there is a `@types/ed2curve`! Removing eslint-disable.

**Comment by RoyTimes** [comment]
I have included test cases for cross sr & ed. I'm fairly positive if the tests can be added and work as expected in a similar fashion as the test for cross sr & ed. However, I'm not sure if it's the right cryptographic thing to do when forcibly converting private keys of the same size from x25519 to secp256k1. Two options here: add in an assert to reject usage on secp256k1 curve; OR just do the conversion and compose test cases for them. What do you think?

**Comment by jacogr** [comment]
As a first-step, rather play safe and assert them away. 

We can re-visit if need be or add an alternative approach for them (which would probably be the correct approach here).

**Comment by RoyTimes** [comment]
Done. 

On a better non-hacky approach: I have no doubt that all of ed25519, sr25519 and secp256k1 can do ECDH; on a worse side, the library for secp256k1 supports ECDH but not encryption/decryption and I might have to implement a portion of the cryptographic myself. On a better side, introduce new dependencies that is well trusted to do ed25519 sr25519 ECDH. Tweetnacl does not have it built in. For sr25519, that will involve messing with the Rust implementation. 

Long story short, it's a pretty heavy load of work to be done. I'm happy to revisit this when I'm more available.

When this is merged, I'll jump to `extension`.

File: packages/keyring/src/pair/index.ts [hunk 5]
```diff
@@ -145,6 +145,18 @@ export function createPair ({ toSS58, type }: Setup, { publicKey, secretKey }: P
     },
     // eslint-disable-next-line sort-keys
     decodePkcs8,
+    decryptMessage: (encryptedMessageWithNonce: string | Uint8Array, senderPublicKey: string | Uint8Array): Uint8Array | null => {
+      assert(!isLocked(secretKey), 'Cannot encrypt with a locked key pair');
+      assert(type !== 'ecdsa', 'Secp256k1 not supported yet');
+      assert(type !== 'ethereum', 'Secp256k1 not supported yet');
```
Comment (jacogr) [inline packages/keyring/src/pair/index.ts:151]:
```suggestion
      assert(!['ecdsa', 'ethereum'].includes(type), 'Secp256k1 not supported yet');
```

File: packages/keyring/src/pair/index.ts [hunk 6]
```diff
@@ -157,6 +169,16 @@ export function createPair ({ toSS58, type }: Setup, { publicKey, secretKey }: P
     encodePkcs8: (passphrase?: string): Uint8Array => {
       return recode(passphrase);
     },
+    encryptMessage: (message: string | Uint8Array, recipientPublicKey: string | Uint8Array, _nonce?: Uint8Array): Uint8Array => {
+      assert(!isLocked(secretKey), 'Cannot encrypt with a locked key pair');
+
+      assert(type !== 'ecdsa', 'Secp256k1 not supported yet');
+      assert(type !== 'ethereum', 'Secp256k1 not supported yet');
```
Comment (jacogr) [inline packages/keyring/src/pair/index.ts:176]:
```suggestion
      assert(!['ecdsa', 'ethereum'].includes(type), 'Secp256k1 not supported yet');
```

**Review by jacogr — APPROVED** [review]
LGTM. That you very much. 

Like mentioned, this is an interesting approach, certainly not something I thought about to get over the sr25519-doesn't-have-that-in-the-libs (yet).
